go/database/mpt/io/parallel_visit.go

```diff
@@ -0,0 +1,462 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package io
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"path"
+	"sync"
+	"sync/atomic"
+
+	"github.com/Fantom-foundation/Carmen/go/backend/stock"
+	"github.com/Fantom-foundation/Carmen/go/backend/stock/file"
+	"github.com/Fantom-foundation/Carmen/go/common/heap"
+	"github.com/Fantom-foundation/Carmen/go/database/mpt"
+)
+
+//go:generate mockgen -source parallel_visit.go -destination parallel_visit_mocks.go -package io
+
+// visitAll visits all nodes in the trie rooted at the given node in depth-first pre-order order.
+// This function accesses nodes using its own read-only node source, independently of a potential
+// node source and cache managed by an MPT Forest instance.
+// Thus, the caller needs to make sure that any concurrently open MPTs have been flushed
+// before calling this function to avoid reading out-dated or inconsistent data.
+func visitAll(
+	directory string,
+	root mpt.NodeId,
+	visitor noResponseNodeVisitor,
+	pruneStorage bool,
+) error {
+	return visitAllSources(&stockNodeSourceFactory{directory}, root, visitor, pruneStorage)
+}
+
+// visitAllSources is an internal implementation of visitAll that allows to provide a custom node source factory.
+func visitAllSources(
```

Comment:
nit: this sounds a bit like sources are visited ... how about `visitAllWithSources`?

Reply:
ok, will change